src/BenchmarkDotNet/Disassemblers/InstructionFormatter.cs

```diff
@@ -0,0 +1,139 @@
+﻿using BenchmarkDotNet.Diagnosers;
+using Iced.Intel;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace BenchmarkDotNet.Disassemblers
+{
+    internal static class CodeFormatter
+    {
+        internal static string Format(SourceCode sourceCode, DisassemblyDiagnoserConfig config, uint pointerSize, IReadOnlyDictionary<ulong, string> addressesMapping)
+        {
+            switch(sourceCode)
+            {
+                case Asm asm:
+                    return InstructionFormatter.Format(asm.Instruction, config, pointerSize, addressesMapping);
+                case Sharp sharp:
+                    return sharp.Text;
+                case MonoCode mono:
+                    return mono.Text;
+                default:
+                    throw new NotSupportedException();
+            }
+        }
+    }
+
+    internal static class InstructionFormatter
+    {
+        internal static string Format(Instruction instruction, DisassemblyDiagnoserConfig config, uint pointerSize, IReadOnlyDictionary<ulong, string> addressesMappings)
+        {
+            var internalBuffer = new StringBuilder(); // we need it to get the length (not exposed by StringBuilderFormatterOutput directly)
+            var output = new StringBuilderFormatterOutput(internalBuffer);
+
+            if (config.PrintInstructionAddresses)
+            {
+                FormatInstructionPointer(instruction, config.Formatter, pointerSize, output);
+            }
+
+            if (ClrMdDisassembler.TryGetReferencedAddress(instruction, pointerSize, out ulong referencedAddress) && addressesMappings.ContainsKey(referencedAddress))
+            {
+                // this instruction refers to an address that we know how to transalte
```

Comment:
Seems like this is code to add symbols to the output? If so you should add your own `ISymbolResolver` and the formatter will call your code to get the symbol. Pass it in to the formatter's constructor. There's rarely a need to call any other method than `Format()`, like the code below in the 'else' block does.

Reply:
thanks for a great hint! I've switched to `ISymbolSolver` now https://github.com/dotnet/BenchmarkDotNet/pull/1332/commits/d14e354d477ae6cf00b57db668dbfce88493d545